Add tuple input support to all-gather and reduce-scatter

jeffhataws opened the pull request:
This PR adds tuple input support to all-gather and reduce-scatter. This is a revival of part of https://github.com/tensorflow/tensorflow/pull/58377 and to be used in conjunction with https://github.com/pytorch/xla/pull/5624 .

In FSDP, different layers' weights need to be all-gathered/reduced-scatter during training. If some layers are small, multiple layers' weights can be aggregated for more efficient data transfer (same concept as bucket_cap_mb in DDP). With existing all-gather and reduce-scatter in PyTorch-XLA, you would have to do the bucketing and decomposing outside of the operation. This PR enables multiple different tensors to be all-gathered/reduce-scatter, keeping the original tensor shapes to enable bucketing and decomposing optimizations inside the operation.

Original PR has token support like the token used for allreduce to ensure order between CCops. That will be separate PR if needed.

radhakrishnaba commented:
Hi @jeffhataws, can you sign up on this CLA provided below.
https://cla.developers.google.com/clas

jeffhataws commented:
> Hi @jeffhataws, can you sign up on this CLA provided below. https://cla.developers.google.com/clas

Done in the latest merge from main.

jurahul commented on xla/client/xla_builder.cc:
```diff
@@ -2883,10 +2883,30 @@ XlaOp XlaBuilder::AllGatherImpl(const XlaOp operand,
+        return Unimplemented("0 element tuple AllGather is not supported");
+      }
+      for (int i = 0; i < operand_shape->tuple_shapes_size(); ++i) {
+        if (operand_shape->tuple_shapes(i).element_type() !=
```
why is this required here? I don't think an all-gather in particular has this restriction at HLO level. For all-reduce such restriction makes sense since the reduction computation is typed.

Also, can we share the tuple flattening code with all-reduce/reduce-scatter and make a helper function for that?

jeffhataws replied:
@jurahul you are right. Removed the check for same shape.

The other request of sharing tuple flattening code will be a larger refactor across existing  all-reduce/reduce-scatter so I want to avoid that to keep the changes to minimum and self-contained.

jurahul reviewed (changes requested):
Overall, I think my comment on the earlier version of this change is still valid. Its doing 3 things (1) adding a convenience to pass in tuples into AllGather and ReduceScatter XLA Builder API and have the API flatten these tuples. and (2) Adding a way to control the ordering of the collective ops by treading them through "tokens", and (3) allowing these threading token to be either a token type in XLA or any scalar type.

Part 1 can be easily split into its own change is not concerning, so that should be the first step. Parts 2 and 3 really need to be discussed with the broader team. Can we start an RFC around this. XLA already has control dependency mechanism to arbitrarily order instructions. This will be a new mechanism for collectives. It seem we should have a common mechanism to express any scheduling constrains as HLO comes into XLA as I can image this token based scheme potentially applicable to all other ops, for various reasons (like controlling memory pressure or something else).

Also, allowing any scalar as a token for these instructions seems confusing and that should be discussed as well. Also, there is another aspect of this needing equivalent changes in StabbleHLO as well.

@burmako can you please advice on how HLO/StableHLO level changes like parts of this PR should be driven?

burmako commented:
+1 to @jurahul's feedback. I would also like to request an RFC for (2) and (3), so that we can discuss design tradeoffs across multiple groups and end up with clear documentation of the new functionality.

We haven't yet standardized on an RFC process for StableHLO, and I've seen several ways of approaching this. My recommendation would be to post a proposal on [openxla-discuss](https://groups.google.com/a/openxla.org/g/openxla-discuss/), with a description of updated semantics for the affected ops. 

Here's a question which may help to get the RFC started. The StableHLO specification has a description of semantics and constraints for the existing collective ops, e.g. here's [an example for all_gather](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#all_gather). How would this specification need to change to accommodate the proposal in this PR?

jeffhataws commented:
> burmako

Thanks @jurahul  and @burmako . Will make modifications based on your feedbacks.

radhakrishnaba commented:
Hi @jeffhataws, any update on this? Thanks.

jeffhataws commented:
> Hi @jeffhataws, any update on this? Thanks.

Thanks for checking. Will post an update by next week, with RFC as requested.

jeffhataws commented:
@jurahul, @burmako, @radhakrishnaba please take a look. I addressed all concerns except the request to refactor the tuple flattening which I thought can be done separately as it touches other parts not related to this PR. The token part has been removed from this PR (maybe a separate PR with RFC) so it should be ready for merge now. Thanks!

jeffhataws commented on xla/service/hlo_verifier.cc:
```diff
@@ -434,6 +434,7 @@ static Status CheckCommonAllGatherInvariants(HloInstruction* hlo,
                                                ag->use_global_device_ids()));
   TF_RETURN_IF_ERROR(CheckReplicaGroups(ag, group_mode));
   TF_RET_CHECK(ag->all_gather_dimension() >= 0);
+  TF_RET_CHECK(ag->operand_count() >= 1);
```
@jurahul, @burmako, @radhakrishnaba I think this is correct. Let me know if you want me to remove this.

jurahul replied:
Also seems ok to me.

jeffhataws commented on xla/service/hlo_verifier.cc:
```diff
@@ -521,6 +522,7 @@ Status ShapeVerifier::HandleReduceScatter(HloInstruction* hlo) {
                                                ars->use_global_device_ids()));
   TF_RETURN_IF_ERROR(CheckReplicaGroups(ars, group_mode));
   TF_RET_CHECK(ars->scatter_dimension() >= 0);
+  TF_RET_CHECK(ars->operand_count() >= 1);
```
same here.

jeffhataws commented on xla/service/cpu/ir_emitter.cc:
```diff
@@ -1289,6 +1289,10 @@ Status IrEmitter::HandleAllReduce(HloInstruction* crs) {
 }
 
+Status IrEmitter::HandleReduceScatter(HloInstruction* rs) {
+  return Unimplemented("ReduceScatter is not implemented on CPU.");
```
@jurahul, @burmako, @radhakrishnaba I think this is correct. Let me know if you want me to remove this.

jurahul replied:
This seems fine to me,

jurahul reviewed (approved):
LGTM

jeffhataws commented:
Hi @thomasjoerg will you help complete the review? Thanks.

GleasonK commented on xla/client/xla_builder_test.cc:
```diff
@@ -418,6 +418,22 @@ TEST_F(XlaBuilderTest, AllGatherR2) {
+  XlaBuilder b(TestName());
+  auto x = Parameter(&b, 0, ShapeUtil::MakeShape(F32, {4}), "x");
+  auto x2 = Parameter(&b, 1, ShapeUtil::MakeShape(F32, {16, 4}), "x2");
+  AllGather(Tuple(&b, {x, x2}), /*all_gather_dimension=*/0,
```
From an API perspective, XlaBuilder has:

```
  XlaOp AllToAll(XlaOp operand, int64_t split_dimension,
                 int64_t concat_dimension, int64_t split_count,
                 absl::Span<const ReplicaGroup> replica_groups,
                 const std::optional<Layout>& layout = std::nullopt,
                 const std::optional<ChannelHandle>& channel_id = std::nullopt);

  XlaOp AllToAllTuple(
      absl::Span<const XlaOp> operands,
      absl::Span<const ReplicaGroup> replica_groups,
      const std::optional<Layout>& layout,
      const std::optional<ChannelHandle>& channel_id = std::nullopt);

XlaOp AllReduce(XlaOp operand, const XlaComputation& computation,
                absl::Span<const ReplicaGroup> replica_groups = {},
                const std::optional<ChannelHandle>& channel_id = std::nullopt,
                const std::optional<Shape>& shape_with_layout = std::nullopt,
                std::optional<bool> use_global_device_ids = std::nullopt);

XlaOp AllReduceTuple(
    absl::Span<const XlaOp> operand, const XlaComputation& computation,
    absl::Span<const ReplicaGroup> replica_groups = {},
    const std::optional<ChannelHandle>& channel_id = std::nullopt,
    const std::optional<Shape>& shape_with_layout = std::nullopt,
    std::optional<bool> use_global_device_ids = std::nullopt);
```

Consider adding an `AllGatherTuple` API for consistency, and using it in this test. See AllReduceTuple impl, it just wraps in tuple and dispatches.

jeffhataws replied:
Thanks. I think it makes sense to make it consistent. I will take a look at this.

JackCaoG commented:
@cheshire @burmako Do you guys know what's the process of merging this pr on the open source side? I saw that copybara is currently failing for this pr.

jurahul commented:
@ddunl can you please help take a look why this is blocked?

ddunl commented:
Disabled the already failing tests by editing `TAP_PROJECTS` in the description, hopefully will submit this upcoming try
